modules/light-clients/08-wasm/keeper/msg_server_test.go

```diff
@@ -88,11 +97,146 @@ func (suite *KeeperTestSuite) TestMsgStoreCode() {
 					suite.Require().Contains(events, evt)
 				}
 			} else {
-				suite.Require().Error(err)
 				suite.Require().ErrorIs(err, tc.expError)
 				suite.Require().Nil(res)
 				suite.Require().Empty(events)
 			}
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestMsgMigrateContract() {
+	oldCodeHash := sha256.Sum256(wasmtesting.Code)
+
+	newByteCode := []byte("MockByteCode-TestMsgMigrateContract")
+
+	govAcc := authtypes.NewModuleAddress(govtypes.ModuleName).String()
+
+	var (
+		newCodeHash    []byte
+		msg            *types.MsgMigrateContract
+		expClientState *types.ClientState
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {
+				msg = types.NewMsgMigrateContract(govAcc, defaultWasmClientID, newCodeHash, []byte("{}"))
+
+				suite.mockVM.MigrateFn = func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, _ wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					data, err := json.Marshal(types.EmptyResult{})
+					suite.Require().NoError(err)
+
+					return &wasmvmtypes.Response{Data: data}, wasmtesting.DefaultGasUsed, nil
+				}
+			},
+			nil,
+		},
+		{
+			"success: update state",
+			func() {
+				msg = types.NewMsgMigrateContract(govAcc, defaultWasmClientID, newCodeHash, []byte("{}"))
+
+				suite.mockVM.MigrateFn = func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					// the code hash written in here will be overwritten
+					expClientState = types.NewClientState([]byte{1}, []byte(ibctesting.InvalidID), clienttypes.NewHeight(2000, 2))
+					store.Set(host.ClientStateKey(), clienttypes.MustMarshalClientState(suite.chainA.App.AppCodec(), expClientState))
+
+					data, err := json.Marshal(types.EmptyResult{})
+					suite.Require().NoError(err)
+
+					return &wasmvmtypes.Response{Data: data}, wasmtesting.DefaultGasUsed, nil
+				}
+			},
+			nil,
+		},
+		{
+			"failure: same code hash",
+			func() {
+				newCodeHash = oldCodeHash[:]
+				msg = types.NewMsgMigrateContract(govAcc, defaultWasmClientID, newCodeHash, []byte("{}"))
+
```

Comment:
does `newCodeHash` need to be reassigned? I find it easier to read if oldCodeHash is provided as arg to msg migrate

Reply:
done